### PR TITLE
[cuDNN][cuDNN V8 API] Thread safety fixes for cuDNN V8 API

### PR DESCRIPTION
**Pull request by eqy**
(this these two fixes are now oudated, see EDIT below)
Fixes for two thread safety issues (one currently unobserved, and one currently observed).
1. `std::erase` can potentially invalidate a pointer to an `ExecutionPlan` in the current implementation. While failures due to this issue have not yet been reported to my knowledge, it is better to return a copy of an `ExecutionPlan` for safety.
2. #103793 surfaced that `cudnnBackendExecute` appears to currently be thread-unsafe. I've verified this with a PyTorch free (pure C++) repro using the cuDNN frontend. This PR addes a mutex that we can hopefully once this issue is resolved.

EDIT:
Feedback from cuDNN is that the V8 backend API has known thread-safety limitations when `ExecutionPlan`s are shared (or even shallow copied) across threads. Given that the common PyTorch use case of eager-mode is singlethreaded (per GPU), this PR now opts to make `ExecutionPlan` caches `thread_local`, as this simplifies the code and eliminates the need for a mutex. The potential tradeoff is some additional warmup cost in multithreaded case, but this would also only be worse than the current case if multiple threads had largely overlapping workloads.

CC @tuero @ptrblck @malfet 

cc @csarofeen @ptrblck @xwang233

### CODE REVIEW

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 1]
```diff
@@ -162,27 +161,27 @@ struct BenchmarkCache {
 std::mutex mutex;
```
Comment (colesbury) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:161]: It looks like the mutex is not used at all now. Can the field in BenchmarkCache be removed?

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 2]
```diff
@@ -162,27 +161,27 @@ struct BenchmarkCache {
...
   if (it == engine_cache.end()) {
     return nullptr;
   }
-  // TODO: probably want ExecutionPlan copy constructor or better way to return
   return &(it->second);
 }
 
 void update(const KeyType& key, T& results) {
-  std::lock_guard<std::mutex> guard(mutex);
   engine_cache.erase(key);
```
Comment (malfet) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:175]: Would this auto release `cudnn_frontend::ExecutionPlan`?

File: aten/src/ATen/native/cudnn/Conv_v8.cpp (hunk 1 above)
Comment (eqy) [reply aten/src/ATen/native/cudnn/Conv_v8.cpp:161]: Yes, thanks!

File: aten/src/ATen/native/cudnn/Conv_v8.cpp (hunk 2 above)
Comment (eqy) [reply aten/src/ATen/native/cudnn/Conv_v8.cpp:175]: Do you mean release in terms of thread safety, or something else (memory management)?

In this case by making the caches `thread_local` each thread would have its own execution plans, even if the workload is the same.

If you mean memory management, then I believe it should call the destructor and then it's cuDNN's responsibility to ensure that the `ExecutionPlan` is properly deallocated. In general, we would expect `erase` on a key that is already in the cache to be rare---it should only happen when an OOM due to a workspace is hit.

**Review by xw285cornell — APPROVED** [review]

**Review by colesbury — APPROVED** [review]
LGTM, thanks @eqy!

**Pull request closed without merging**